getComputedStyle() returns the right values for invalid pseudoElement

karlcow opened the pull request:
#### c7fe66d121e6466441ec0d8d15b8c5258462fa11
<pre>
getComputedStyle() returns the right values for invalid pseudoElement
<a href="https://bugs.webkit.org/show_bug.cgi?id=243539">https://bugs.webkit.org/show_bug.cgi?id=243539</a>
<a href="https://rdar.apple.com/98504661">rdar://98504661</a>

Reviewed by NOBODY (OOPS!).

FIXME: Needs to change the description and list of files as
<a href="https://github.com/WebKit/WebKit/pull/22231">https://github.com/WebKit/WebKit/pull/22231</a>
fixed some of the issues in a more general way than the way this
patch was trying to do. Or maybe I should just closed this PR and
starts fresh.

parseStandalonePseudoElement() should probably be moved in
CSSComputedStyleDeclaration.cpp

The getComputedStyle(elt, pseudoElt) method must run these steps:

1. Let doc be elt&apos;s node document.
2. Let obj be elt.
3. If pseudoElt is provided, is not the empty string, and starts with a colon, then:
   3.1 Parse pseudoElt as a &lt;pseudo-element-selector&gt;, and let type be the result.
   3.2 If type is failure, or is an ::slotted() or ::part() pseudo-element, let obj be null.
   3.3 Otherwise let obj be the given pseudo-element of elt.
       Note: CSS2 pseudo-elements should match both the double
       and single-colon versions. That is, both :before
       and ::before should match above.
4. Let decls be an empty list of CSS declarations.
5. If obj is not null, and elt is connected, part of the flat tree, and its shadow-including root has a browsing context which either doesn’t have a browsing context container, or whose browsing context container is being rendered, set decls to a list of all longhand properties that are supported CSS properties, in lexicographical order, with the value being the resolved value computed for obj using the style rules associated with doc. Additionally, append to decls all the custom properties whose computed value for obj is not the guaranteed-invalid value.

* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::CSSComputedStyleDeclaration):
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::pseudoId):
(WebCore::CSSSelector::parsePseudoElement):
(WebCore::CSSSelector::parseStandalonePseudoElement):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c7fe66d121e6466441ec0d8d15b8c5258462fa11

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31935 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10632 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33687 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34448 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/28931 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32737 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12998 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7870 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28526 "Found 1 new test failure: fast/css/getComputedStyle/getComputedStyle-with-pseudo-element.html (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32304 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8991 "Found 1 new test failure: fast/css/getComputedStyle/getComputedStyle-with-pseudo-element.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28536 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7776 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7957 "Exiting early after 10 failures. 20 tests run. 1 failures") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28449 "Found 1 new test failure: fast/css/getComputedStyle/getComputedStyle-with-pseudo-element.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35792 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29053 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28903 "Found 2 new test failures: fast/css/getComputedStyle/getComputedStyle-with-pseudo-element.html, tiled-drawing/scrolling/scroll-snap/scroll-snap-mandatory-mainframe-vertical-then-horizontal.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34055 "Found 1 new test failure: fast/css/getComputedStyle/getComputedStyle-with-pseudo-element.html (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8041 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6027 "Found 29 new test failures: fast/css/getComputedStyle/getComputedStyle-with-pseudo-element.html, fast/media/media-query-dynamic-with-font-face.html, inspector/css/add-css-property.html, inspector/css/css-property.html, inspector/css/forcePseudoState.html, inspector/css/generateFormattedText.html, inspector/css/getComputedPrimaryFontForNode.html, inspector/css/getMatchedStylesForNode.html, inspector/css/getMatchedStylesForNodeBackdropPseudoId.html, inspector/css/getMatchedStylesForNodeContainerGrouping.html ... (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31914 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9687 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/28256 "Exiting early after 10 failures. 20 tests run. 1 failures") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8704 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8564 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

karlcow commented on Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
```diff
@@ -51,10 +51,21 @@ CSSComputedStyleDeclaration::CSSComputedStyleDeclaration(Element& element, bool
+    // handle normal case "::marker" and "::foobar"
+    else
+        name = name.substring(2);
+
```
@anttijk This fixes the parsing issue but I'm not sure yet how to solve the getComputedStyle returned value (also the comment are here to explain my choice, I can remove them later if necessary).
What do you think should be done? Or should it be a separate issue?

karlcow replied:
Context for the returned values in 
https://github.com/web-platform-tests/wpt/pull/37452#discussion_r1046510669

nt1m replied:
As mentioned on Slack, returning nullptr here: https://github.com/WebKit/WebKit/blob/8d03c66477e27515ed83a15d6e9f66ec484f2898/Source/WebCore/css/ComputedStyleExtractor.cpp#L2490-L2491 will probably yield the correct return value.

That might be done through nulling out `m_element` or adding a new PseudoId::Invalid and handling it in that method.

anttijk replied:
You could do
`std::optional<PseudoId> m_pseudoElementSpecifier;`
and where nullopt would signal that there is no pseudo element and PseudoId::None would signal that there was but it failed to parse.

Or if that is confusing then you could just have a separate parse failure bit.

karlcow replied:
@nt1m 
I tried to find a way to nullify the element but without success.
I managed a point to have the desired result by touching Element.cpp, but I clearly nullified the style for the parent element itself. 😮‍💨 

So I need a bit of help here. 

Does `m_element` can be modified? or is it constant?

And where could I do something like `m_element = nullptr;`


The parsing is all good. This is just the last stretch where I"m hitting the wall.

karlcow committed:
getComputedStyle() returns the right values for invalid pseudoElement
https://bugs.webkit.org/show_bug.cgi?id=243539
rdar://98504661

Reviewed by NOBODY (OOPS!).

FIXME: Needs to change the description and list of files as
https://github.com/WebKit/WebKit/pull/22231
fixed some of the issues in a more general way than the way this
patch was trying to do. Or maybe I should just closed this PR and
starts fresh.

parseStandalonePseudoElement() should probably be moved in
CSSComputedStyleDeclaration.cpp

The getComputedStyle(elt, pseudoElt) method must run these steps:

1. Let doc be elt's node document.
2. Let obj be elt.
3. If pseudoElt is provided, is not the empty string, and starts with a colon, then:
   3.1 Parse pseudoElt as a <pseudo-element-selector>, and let type be the result.
   3.2 If type is failure, or is an ::slotted() or ::part() pseudo-element, let obj be null.
   3.3 Otherwise let obj be the given pseudo-element of elt.
       Note: CSS2 pseudo-elements should match both the double
       and single-colon versions. That is, both :before
       and ::before should match above.
4. Let decls be an empty list of CSS declarations.
5. If obj is not null, and elt is connected, part of the flat tree, and its shadow-including root has a browsing context which either doesn’t have a browsing context container, or whose browsing context container is being rendered, set decls to a list of all longhand properties that are supported CSS properties, in lexicographical order, with the value being the resolved value computed for obj using the style rules associated with doc. Additionally, append to decls all the custom properties whose computed value for obj is not the guaranteed-invalid value.

* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::CSSComputedStyleDeclaration):
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::pseudoId):
(WebCore::CSSSelector::parsePseudoElement):
(WebCore::CSSSelector::parseStandalonePseudoElement):
